contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -73,153 +76,211 @@ contract LidoOracle is ILidoOracle, AragonApp {
-        if (index != last) {
-            members[index] = members[last];
-            bitMask = bitMask.setBit(index, bitMask.getBit(last));
-        }
-        bitMask = bitMask.setBit(last, false);
-        gatheredEpochData[lastReportedEpochId].reportsBitMask = bitMask;
-
+        if (index != last) members[index] = members[last];
         members.length--;
-
         emit MemberRemoved(_member);
-        _assertInvariants();
+
+        // Nullify the data for the last epoch. Remained oracles will report it again
+        REPORTS_BITMASK_POSITION.setStorageUint256(0);
+        delete gatheredReportKinds;
     }
 
     /**
      * @notice Set the number of oracle members required to form a data point to `_quorum`
      */
     function setQuorum(uint256 _quorum) external auth(MANAGE_QUORUM) {
-        require(members.length >= _quorum && 0 != _quorum, "QUORUM_WONT_BE_MADE");
-
+        require(0 != _quorum, "QUORUM_WONT_BE_MADE");
         QUORUM_POSITION.setStorageUint256(_quorum);
         emit QuorumChanged(_quorum);
+        _tryPush(REPORTABLE_EPOCH_ID_POSITION.getStorageUint256());
+    }
 
-        uint256 minReportableEpochId = (
-            MIN_REPORTABLE_EPOCH_ID_POSITION.getStorageUint256()
-        );
-        uint256 lastReportedEpochId = (
-            LAST_REPORTED_EPOCH_ID_POSITION.getStorageUint256()
-        );
-
-        assert(lastReportedEpochId <= getCurrentEpochId());
-
-        if (lastReportedEpochId > minReportableEpochId) {
-            minReportableEpochId = lastReportedEpochId;
-            _tryPush(lastReportedEpochId);
-        }
+    /**
+     * @notice Returns the callback contract address to be called upon quorum
+     */
+    function getQuorumCallback() public view returns(address) {
+        return address(QUORUM_CALLBACK_POSITION.getStorageUint256());
+    }
 
-        _assertInvariants();
+    /**
+     * @notice Set the callback contract address to be called upon quorum
+     */
+    function setQuorumCallback(address _addr) external auth(SET_QUORUM_CALLBACK) {
+        QUORUM_CALLBACK_POSITION.setStorageUint256(uint256(_addr));
+        emit QuorumCallbackSet(_addr);
     }
 
     /**
      * @notice An oracle committee member reports data from the ETH 2.0 side
-     * @param _epochId BeaconChain epoch id
+     * @param _epochId Beacon Chain epoch id
      * @param _beaconBalance Balance in wei on the ETH 2.0 side
      * @param _beaconValidators Number of validators visible on this epoch
      */
     function reportBeacon(uint256 _epochId, uint128 _beaconBalance, uint128 _beaconValidators) external {
-        (uint256 startEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
-        require(_epochId >= startEpoch, "EPOCH_IS_TOO_OLD");
+        BeaconSpec memory beaconSpec = _getBeaconSpec();
+        (uint256 currentEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
+        require(_epochId >= currentEpoch, "EPOCH_IS_TOO_OLD");
```

Comment:
I propose to prohibit accepting any epochs other than the current frame. Why are we allowing the range here?

Reply:
Fix in https://github.com/lidofinance/lido-dao/pull/311/files for clarity.